### ISSUE TITLE
Using ImplicitlyWait

### ISSUE THREAD

**Issue by tarurar** [Bumblebee/Bumblebee#90]
Hi, I've tried to use **ImplicitlyWait** because there are a lot of content on my site generated by javascript. Below is the example of the unit test. The scenario is simple: 
1. Opening list view page
2. Clicking New Item link (after that new item page should open)
3. Filling title field with some random value
4. Clicking Save button (after that new item page should close and list view page should open again)
5. Checking if we are on the list view page indeed

``` c#
[TestFixture]
public class NewItemTests
{
    protected Session Session { get; private set; }

    [SetUp]
    public void SetUp()
    {
        Session = new Session<Chrome>();
        Session.Driver.Manage().Timeouts().ImplicitlyWait(new TimeSpan(0, 0, 10));
    }

    [TearDown]
    public void TearDown()
    {
        Session.End();
    }

    [Test]
    public void CanAddNewItem_Bumblebee()
    {
        var newItemTitle = RandomString();

        Session
            .NavigateTo<SPListViewPage>("<the url of the site>")
            .NewItemLink
            .Click<SPListNewItemPage>()
            .TitleField.EnterText(newItemTitle)
            .SaveRibbonButton.Click<SPListViewPage>()
            .VerifyPresenceOf<SPListViewPage>("new item link", By.Id("idHomePageNewItem"));
    }
}
```

The problem is that every other time the test fails with the error like this:

> Result Message:   OpenQA.Selenium.StaleElementReferenceException : stale element reference: element is not attached to the page document
>   (Session info: chrome=48.0.2564.109)
>   (Driver info: chromedriver=2.21.371459 (36d3d07f660ff2bc1bf28a75d1cdabed0983e7c4),platform=Windows NT 6.3 x86_64)

As you can see even ImplicitlyWait doesn't help.
After that I tried to do the same test but with pure Selenium. Below is the same scenario.

``` c#
[TestFixture]
public class NewItemTests
{
    [Test]
    public void CanAddNewItem_Selenium()
    {
        var newItemTitle = RandomString();

        using (var driver = new ChromeDriver())
        {
            driver.Manage().Timeouts().ImplicitlyWait(new TimeSpan(0, 0, 10));

            driver.Navigate().GoToUrl("<the url of the site>");
            var newItemLink = driver.FindElementById("idHomePageNewItem");
            newItemLink.Click();

            var titleField = driver.FindElementByCssSelector("input[id^=Title]");
            titleField.Clear();
            titleField.SendKeys(newItemTitle);

            var saveButton = driver.FindElementByCssSelector("input[id*='toolBarTbl_'][id*='SaveItem']");
            saveButton.Click();

            Assert.AreEqual(true, driver.FindElementById("idHomePageNewItem").Displayed);

            driver.Quit();
        }
    }
}
```

This code always works well without any issues. Only if I remove the below line it fails.

``` c#
driver.Manage().Timeouts().ImplicitlyWait(new TimeSpan(0, 0, 10));
```

As you can see both fragments are almost identical. So, my guess is that something is wrong with ImplicitlyWait in Bumblebee.

Do you have any thoughts on it?

**Comment by toddmeinershagen** [comment]
A few questions:  

* Which version of Bumblebee are you using?
* I need more info to reproduce the problem in my environment.  Would you provide the link to the site and the code for the SPListViewPage block that you are using?  Or maybe you can put your code in gists or in a github repository so that I can reference it from there.
* Normally, the Session.Driver is not manipulated directly (that is the selenium IWebDriver).  Instead, you can set up the Session by calling the following:

```csharp
var timeToWait = TimeSpan.FromSeconds(10);
Session = new Session(new ChromeDriver(timeToWait));
```

**Comment by tarurar** [comment]
- I am using Bumblebee Automation NuGet package, 1.2.0.0. version
- I've changed the initialization code of the driver but it doesn't help
- Unfortunately, the site is behind the firewall, but I can share the full sources with you. Here is [the link](https://github.com/tarurar/ETEA.Bumblebee.Tests.git).

**Comment by toddmeinershagen** [comment]
Okay.  It will be a little hard to find the problem with the 1.2 framework
without having access to the site, since I can't reproduce the exact issue
that you are experiencing.  The error message, though, indicates that an
element reference is going stale.  The v1 of Bumblebee grabs references to
the Parent tag and the element itself when constructing the web block.  In
the meantime, the page may have shifted dynamically underneath by the time
that an action is called.

The v2 of Bumblebee solves this problem by getting elements right before
calling the action.  So, the locating of elements are just in time (lazy
loaded) at the time of the Click or entry of text, etc.  I recommend that
you try installing the v2 by doing the following at the Package Manager
console:

>Install-Package Bumblebee.Automation -Pre

Let me know if that fixes your issue.

Thanks,



Todd

**Comment by tarurar** [comment]
Hi.
I've updated Bumblebee but result is the same. Please check the attached screenshot.

![bumblebee_2060](https://cloud.githubusercontent.com/assets/142044/13201215/0fb2d784-d881-11e5-86f5-50a1173d05e6.png)

**Comment by toddmeinershagen** [comment]
Okay...thanks for updating your project with the latest Bumblebee and sending the screenshot.  The full stack trace is helpful.

It appears that the problem is in the VerifyPresentOf<T> method.  I noticed that your assertion in the Selenium example is different in a couple of ways.  

1) You are calling FindElementById("idHomePageNewItem") off of the driver, while the Bumblebee verification method is calling FindElementsById("idHomePageNewItem") and checking for .Any().

2) You are calling straight from the Driver whereas the Bumblebee verification method is calling from a scoped body element -  meaning that it has to call var tag = Driver.FindElementByTagName("body") and then call tag.FindElementsById("[input....") in order to find the actual link from the body.

I am wondering if you can try the first option in your Selenium code.  If you call FindElementsById and check for .Any(), do you get an error?

If that does not error out, can you try the second assertion where you get the body element and then get the other element by calling FindElementsById()?  

One of those two should fail, and it will help me to identify the root cause of your issue.

This has also brought to my attention that we are not making the .Displayed property available on the base Element class, which might be useful to users.

Let me know what you find, since I cannot reproduce the issue locally.

**Comment by tarurar** [comment]
Yes, you are right. After the second option applied I've got the same error. Here is a screenshot.
![bumblebee_2060-2](https://cloud.githubusercontent.com/assets/142044/13202653/79da6e4e-d8bb-11e5-804f-6dfd9bbf7e2b.png)

**Comment by toddmeinershagen** [comment]
Okay...so, it appears that it has something to do with scoping.  Can you call FindElement() instead of FindElements off of the bodyTag to see if it is the particular method?  Let me know what you find.

**Comment by tarurar** [comment]
The code was as the following:
```c#
 var bodyTag = driver.FindElementByTagName("body");
 //Assert.AreEqual(true, bodyTag.FindElements(By.Id("idHomePageNewItem")).Any());
 Assert.AreEqual(true, bodyTag.FindElement(By.Id("idHomePageNewItem")).Displayed);
```

... and the error was the same exactly.

**Comment by toddmeinershagen** [comment]
Thanks for your quick response.  This has been really helpful!

Can you paste the screenshot with the stack trace into your next comment?

**Comment by tarurar** [comment]
Do you need the stacktrace from the latest trial?
If so, here it is.
![bumblebee_2060-3](https://cloud.githubusercontent.com/assets/142044/13202770/4b6473de-d8c0-11e5-8e87-bba67256d449.png)

**Comment by toddmeinershagen** [comment]
It looks like the body element has either been deleted or no longer attached to the DOM within the milliseconds it took to call FindElement off of it for the link element.

[This article](http://stackoverflow.com/questions/16166261/selenium-webdriver-stale-element-reference-exception) on StackOverflow explains it like this:

> There is a slim possibility that you will get a StaleElementReference exception with a java PageFactory if you are unlucky enough to have an element go stale in the couple of ms between the PageFactory finding the element, and then using it to do something. – Ardesco Jan 4 at 12:05

[This is an another article](http://docs.seleniumhq.org/exceptions/stale_element_reference.jsp) from the Selenium team explaining the causes of a StaleElementReferenceException.


Can you check to see if the bodyTag is not null and Displayed?  (Just curious if we can confirm that it exists and is not visible.)

**Comment by toddmeinershagen** [comment]
Another thing to check is actually pausing (Thread.Sleep(TimeSpan.FromSeconds(1)) before calling to get the bodyTag....

If that works, then you can actually pass a longer timespan into the WebBlock constructor when you are building your SPListViewPage.  Maybe, TimeSpan.FromSeconds(1) to get it to pause before setting up the Tag.

**Comment by tarurar** [comment]
As for the body tag, I can't confirm that it exists and not visible. I've been running the following test a lot of times:
```c#
var bodyTag = driver.FindElementByTagName("body");
Assert.NotNull(bodyTag);
Assert.AreEqual(true, bodyTag.Displayed);
```
... but always it was successfully completed.

> Another thing to check is actually pausing (Thread.Sleep(TimeSpan.FromSeconds(1)) before calling to get the bodyTag....Try that in your selenium code to see if it will address the issue as well.

Yes, this works well. Each time test passes successfully. 

> If that works, then you can actually pass a longer timespan into the WebBlock constructor when you are building your SPListViewPage

What do you actually mean? I am not building the instance of the SPListViewPage, I am not calling the constructor directly. Besides, in Bumblebee v.2.0.6 the page is not derived from WebBlock, it is now derived from WebPage. So, what should I change to implement your suggestion?

**Comment by toddmeinershagen** [comment]
Actually, in the constructor of your SPListViewPage page you can call Pause(1000).

**Comment by tarurar** [comment]
Got it, but don't you think this is not very good solution?

Also, what in this case means the next way session initialization?
```c#
 Session = new Session(new Chrome(TimeSpan.FromSeconds(10)));
```
Before I thought that this is an equal action because I am telling the driver to wait. Can you clarify?

**Comment by toddmeinershagen** [comment]
Setting the implicit wait is different from pausing until trying to find
the element.

If you set the implicit wait, that means that the framework will look for
an element from the beginning of when you access that element until either
it finds it or the timeout occurs.

In the case of Pause, you are actually waiting for a given time period
before beginning to look for something.  In this case, you are waiting for
the page to settle down, because it gets changed after the initial load and
gets stale, so if you get the reference too early, it looks like it gets
changed or removed from the DOM.  Another option is to set the Tag in the
constructor of your page to some <div> element or high level scope that
would perhaps appear later than the body change.

(Don't know if that last sentence made sense or not...but as the industry
moves toward more dynamic JavaScript and SPA applications, I think we will
see this problem.  The Selenium team knows it is a problem, and you just
have to play with the timings in order to get it right.  No framework can
avoid the problem entirely.)

Hope that helps.

**Comment by toddmeinershagen** [comment]
Also, one option would be to create your own base page that does this wait on every page, if every page is dynamically loading like this.  Then all of your pages can inherit from this page instead of WebPage.  

For instance,

```csharp
public abstract class SPWebPage : WebPage
{
	protected SPWebPage(Session session, TimeSpan timeout) 
		: base(Initialize(session), timeout)
	{}

	protected SPWebPage(Session session) 
		: base(Initialize(session))
	{}

	private Session Initialize(Session session)
	{
		this.Pause(1000);
		return Session;
	}
}
```

**Comment by toddmeinershagen** [comment]
Let me know if all of this makes sense, and if so, I can close the issue.

**Comment by tarurar** [comment]
Thank you

**Comment by tarurar** [comment]
> In the case of Pause, you are actually waiting for a given time period
before beginning to look for something

In my case, every page is dynamically loading and actually the time of loading is going to depend on different circumstances.
For example, if a page loads for the first time after restarting the server it will take much more than 1000ms to load the content. Some pages have heavy logic and this fact also slows down the page loading and initialization. The server itself can be busy and the request processing can take more time than usually. So, the problem is that in some cases even 100ms would be enough while in other cases even 5000ms are going to be a too small period of time.

I know this is not your issue, this is a common issue of the dynamic pages. But it is very annoying problem.

**Comment by Bakanych** [comment]
You can use js techniques to wait for server calls to be returned in your dynamic page instead of using static pause.

**Comment by toddmeinershagen** [comment]
@tarurar I am assuming that you were able to work around your situation of having elements that don't appear immediately.  I think the v2 framework should help in this regard, because we have adding lazy loading of the nested blocks/elements and a wait feature.

I am going to close this for now, but if you are still experiencing issues, feel free to add a new issue and link it to this original issue.
